src/AnalyticsEventReporter.ts

```diff
@@ -1,38 +1,71 @@
-import { AnalyticsEventService } from "./AnalyticsEventService"
-import { AnalyticsConfig } from "./AnalyticsConfig";
-import { EventPayload } from "./EventPayload";
-import { EventAPIResponse } from "./EventAPIResponse";
-import { merge } from "./merge";
+import { AnalyticsEventService } from './AnalyticsEventService';
+import { AnalyticsConfig } from './AnalyticsConfig';
+import { EventPayload, PartialPayload } from './EventPayload';
+import { EventAPIResponse } from './EventAPIResponse';
+import { getOrSetupSessionId } from './setupSessionId';
+import { name, version } from '../package.json';
+import { post } from './post';
+import merge from './merge';
+import { setupRequestUrl } from './setupRequestUrl';
 
 /** Represents an reporter is responsible for reporting analytics events. */
 export class AnalyticsEventReporter implements AnalyticsEventService {
-
     private config: AnalyticsConfig;
     private payload?: EventPayload;
-
     /**
      * @param config - necessary analytics config: Must provide one and only
      * one of API Key or Bearer Token.
-     * 
+     *
      * @param payload - (optional) desired event values to report
      */
     constructor(config: AnalyticsConfig, payload?: EventPayload) {
-        const apiKeyIsSet = config.key !== undefined;
-        const bearerTokenIsSet = config.bearer !== undefined;
-        const configIsValid = (apiKeyIsSet || bearerTokenIsSet) && !(apiKeyIsSet && bearerTokenIsSet);
-        if (!configIsValid) {
-            throw new Error("Provide one and only one of API Key or Bearer Token.")
-        }
-        this.config = config;
-        this.payload = payload;
+      const apiKeyIsSet = config.key !== undefined;
+      const bearerTokenIsSet = config.bearer !== undefined;
+      const configIsValid = (apiKeyIsSet || bearerTokenIsSet) && !(apiKeyIsSet && bearerTokenIsSet);
+      if (!configIsValid) {
+        throw new Error('Provide one and only one of API Key or Bearer Token.');
+      }
+      this.config = config;
+      this.payload = payload;
     }
 
     with(payload: EventPayload): AnalyticsEventService {
-        const currentPayload = this.payload === undefined ? payload : merge(this.payload, payload);
-        return new AnalyticsEventReporter(this.config, currentPayload);
+      const currentPayload = this.payload === undefined ? payload : merge(this.payload, payload);
+      return new AnalyticsEventReporter(this.config, currentPayload);
     }
 
-    report(payload: object): Promise<EventAPIResponse> | boolean {
-        return false;
+    public async report(newPayload?: PartialPayload): Promise<boolean | EventAPIResponse> {
+      const finalPayload = (this.payload && newPayload) ? (merge(this.payload, newPayload))
+        : this.payload ?? newPayload ?? function(){throw Error('EventPayload is empty');}();
+
+      if (this.config.sessionTrackingEnabled) {
```

Comment:
If the user passes in sessionId, it will be added to the finalPayload regardless of if sessionTracking is enabled. 

If it is enabled, we check if it's already been set in the payload and if it has use that value, otherwise call getOrSetupSessionId to get the value. 

So this ensures that if sessionTracking is enabled, then sessionId is set. Otherwise its only set if included in payload.